src/main/scala/org/deepdive/ddlog/DeepDiveLogSemanticChecker.scala

```diff
@@ -142,6 +147,43 @@ object DeepDiveLogSemanticChecker extends DeepDiveLogHandler {
     }
   }
 
+  // check if head variables have bindings in the body
+  def checkHeadVariables(stmt: Statement) {
+    var variableSet = new HashSet[String]()
+    def addVariable(body: List[Body]) {
+      body foreach {
+        case a: Atom => a.terms.foreach {
+          case VarExpr(v) => variableSet += v
+          case _ =>
+        }
+        case a: QuantifiedBody => addVariable(a.bodies)
+        case _ =>
+      }
+    }
+    def checkHeadExpr(e: Expr) {
+      e match {
+        case VarExpr(v) =>
+          if (!variableSet.contains(v)) error(stmt, s"Head variable ${v} is not defined in the body")
+        case FuncExpr(_, args, _) => args foreach checkHeadExpr
+        case BinaryOpExpr(lhs, _, rhs) => {
+          checkHeadExpr(lhs)
+          checkHeadExpr(rhs)
+        }
+        case TypecastExpr(lhs, _) => checkHeadExpr(lhs)
+        case _ =>
+      }
+    }
+    def checkCq(cq: ConjunctiveQuery) {
+      cq.bodies foreach addVariable
+      cq.head.terms foreach checkHeadExpr
+    }
+    stmt match {
+      case s: ExtractionRule => checkCq(s.q)
+      case s: InferenceRule  => checkCq(s.q)
```

Comment:
Weight is checked in `checkWeight `. I'll also add check label, and these functions may be merged into one.